src/main/native/com/code_intelligence/jazzer/android/dex_file_manager.cpp

```diff
@@ -0,0 +1,212 @@
+    std::vector<ir::EncodedField*>* encFieldList) {
+  std::vector<std::string> fieldDescs;
+
+  for (int i = 0; i < encFieldList->size(); i++) {
+    std::stringstream ss;
+    ss << (*encFieldList)[i]->access_flags;
+    ss << (*encFieldList)[i]->decl->type->descriptor->c_str();
+    ss << (*encFieldList)[i]->decl->name->c_str();
+    fieldDescs.push_back(ss.str());
+  }
+
+  sort(fieldDescs.begin(), fieldDescs.end());
+  return fieldDescs;
+}
+
+bool matchFields(std::vector<ir::EncodedField*>* encodedFieldListOne,
+                 std::vector<ir::EncodedField*>* encodedFieldListTwo) {
+  std::vector<std::string> fDescListOne =
+      getFieldDescriptions(encodedFieldListOne);
+  std::vector<std::string> fDescListTwo =
+      getFieldDescriptions(encodedFieldListTwo);
+
+  if (fDescListOne.size() != fDescListTwo.size()) {
+    return false;
+  }
+
+  for (int i = 0; i < fDescListOne.size(); i++) {
+    if (strcmp(fDescListOne[i].c_str(), fDescListTwo[i].c_str()) != 0) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool matchMethods(std::vector<ir::EncodedMethod*>* encodedMethodListOne,
+                  std::vector<ir::EncodedMethod*>* encodedMethodListTwo) {
+  std::vector<std::string> mDescListOne =
+      getMethodDescriptions(encodedMethodListOne);
+  std::vector<std::string> mDescListTwo =
+      getMethodDescriptions(encodedMethodListTwo);
+
+  if (mDescListOne.size() != mDescListTwo.size()) {
+    return false;
+  }
+
+  for (int i = 0; i < mDescListOne.size(); i++) {
+    if (strcmp(mDescListOne[i].c_str(), mDescListTwo[i].c_str()) != 0) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool classStructureMatches(ir::Class* classOne, ir::Class* classTwo) {
+  bool result = true;
+  result = result && matchMethods(&(classOne->direct_methods),
+                                  &(classTwo->direct_methods));
+  result = result && matchMethods(&(classOne->virtual_methods),
+                                  &(classTwo->virtual_methods));
+  result = result &&
+           matchFields(&(classOne->static_fields), &(classTwo->static_fields));
+  result = result && matchFields(&(classOne->instance_fields),
+                                 &(classTwo->instance_fields));
+  result = result && (classOne->access_flags == classTwo->access_flags);
+
+  return result;
+}
+
+bool DexFileManager::structureMatches(dex::Reader* oldReader,
+                                      dex::Reader* newReader,
+                                      const char* className) {
+  std::string descName = GetName(className);
+
+  dex::u4 oldReaderIndex = oldReader->FindClassIndex(descName.c_str());
+  dex::u4 newReaderIndex = newReader->FindClassIndex(descName.c_str());
+
+  if (!IsValidIndex(oldReaderIndex) || !IsValidIndex(newReaderIndex)) {
+    return false;
+  }
+
+  oldReader->CreateClassIr(oldReaderIndex);
+  newReader->CreateClassIr(newReaderIndex);
+
+  std::shared_ptr<ir::DexFile> oldDexFile = oldReader->GetIr();
```

Comment:
Could this be an `std::unique_ptr` instead for simplicity?

Reply:
std::shared_ptr is returned from Slicer, I think it makes sense to not try to convert it so unique_ptr, and also to leave the existing function as is in Slicer.